File: 6. Zigzag Conversion/convert.py
class Solution:
  def convert(self, s: str, numRows: int) -> str:
    if numRows == 1 or numRows >= len(s):
      return s

    rows = [''] * numRows
    row = 0
    going_up = True

    for i in range(len(s)):
      rows[row] += s[i]
      if row == 0 or (going_up and row < numRows - 1):
        row += 1
        going_up = True
      else:
        row -= 1
        going_up = False

    return ''.join(rows)


s = "PAYPALISHIRING"
numRows = 3
solution = Solution()
print('convert: ', solution.convert(s, numRows))

File: 6. Zigzag Conversion/convert.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
const convert = (s, numRows) => {
  if (numRows === 1 || numRows >= s.length) {
    return s;
  }

  const rows = Array(numRows).fill('');
  let row = 0;
  let goingUp = true;

  for (let i = 0; i < s.length; i++) {
    rows[row] += s[i];
    if (row === 0 || (goingUp && row < numRows - 1)) {
      row++;
      goingUp = true
    } else {
      row--;
      goingUp = false;
    }
  }

  return rows.join('');
};


const s = "PAYPALISHIRING";
const numRows = 3;
console.log('convert: ', convert(s, numRows));
